File: src/entities/account.py
from src.models.entities import AccountModel
from src.utils.general import AccountErrors

class Account:
    def __init__(self, number: str, account_type: str, balance: float, special_credit: float = 0.0):
        self.__number = number
        self.__account_type = account_type
        self.__balance = balance
        self.__special_credit = special_credit
        self.owner = None
    
    def get_special_credit(self):
        return self.__special_credit

    def get_balance(self):
        if self.__have_owner:
            return self.__balance
    
    def get_details(self) ->  AccountModel:
        if self.__have_owner:
            return AccountModel(
            number=self.__number,
            balance=self.__balance,
            type = self.__account_type,
            owner=self.owner
            )

    def set_owner(self, person_identity: str):
        self.owner = person_identity
        
    def update_balance(self, amount: float):
        if self.__balance < 0 and amount < 0:
            self.__balance += amount
        else:
            self.__balance += amount
    
    @property
    def __have_owner(self):
        if self.owner is None:
            raise AccountErrors.NO_OWNER.value
        return True

File: go/pkg/entities/account.go
package entities

import (
	"fmt"
)

type Account struct {
	Number        string
	AccountType   string
	Balance       float32
	Owner         string
	SpecialCredit float32
}

type IAccount interface {
	AccountNumber() string
	GetOwner() string
	Type() string
	GetSpecialCredit() float32
	SetSpecialCredit(limit float32) error
	GetBalance() float32
	SetOwner(identity string) error
	UpdateBalance(amount float32)
	haveOwner() bool
}

func NewAccount(number, accountType string, balance float32) (*Account, error) {
	if number == "" || accountType == "" {
		return &Account{}, fmt.Errorf("You need to provide number and account type")
	}
	return &Account{
		Number:        number,
		AccountType:   accountType,
		Balance:       balance,
		SpecialCredit: 0.0,
	}, nil
}

func (a *Account) GetOwner() string {
	return a.Owner
}

func (a *Account) GetSpecialCredit() float32 {
	return a.SpecialCredit
}

func (a *Account) AccountNumber() string {
	return a.Number
}

func (a *Account) Type() string {
	return a.AccountType
}

func (a *Account) SetSpecialCredit(limit float32) error {
	if limit == 0 {
		return fmt.Errorf("You provided a invalid new limit")
	}
	a.SpecialCredit = +limit
	return nil
}

func (a *Account) GetBalance() float32 {
	if a.haveOwner() {
		return a.Balance
	}
	return 0.0
}

func (a *Account) SetOwner(identity string) error {
	if identity == "" {
		return fmt.Errorf("You provided a empty owner")
	}
	a.Owner = identity
	return nil
}

func (a *Account) UpdateBalance(amount float32) {
	if a.haveOwner() {
		if a.Balance < float32(0) {
			if amount > float32(0) {
				a.Balance += amount
			} else {
				a.Balance -= amount
			}
		} else if a.Balance == float32(0) {
			if amount > float32(0) {
				a.Balance += (amount * -1)
			} else {
				a.Balance += amount
			}
		} else {
			a.Balance += amount
		}
	}
}

func (a *Account) haveOwner() bool {
	if a.Owner != "" {
		return true
	} else {
		return false
	}
}
